hwcomposer/hwcomposer_backend_v11.cpp

```diff
@@ -86,97 +94,415 @@ static void hwc11_callback_invalidate(const struct hwc_procs *)
+        uint32_t configs[numConfigs];
+        if (hwc_device->getDisplayConfigs(hwc_device, dpy, configs, &numConfigs) != 0) {
+            fprintf(stderr, "Display %d not connected, no configs\n", dpy);
+            continue;
+        }
 
-    QPA_HWC_TIMING_SAMPLE(presentTime);
+        fprintf(stderr, "%d configs found for display %d\n", numConfigs, dpy);
 
-    fblayer->handle = buffer->handle;
-    fblayer->releaseFenceFd = -1;
+        for (uint i=0; i<numConfigs; i++) {
+            uint32_t attributes[] = {
+                HWC_DISPLAY_VSYNC_PERIOD,
+                HWC_DISPLAY_WIDTH,
+                HWC_DISPLAY_HEIGHT,
+                HWC_DISPLAY_DPI_X,
+                HWC_DISPLAY_DPI_Y,
+                HWC_DISPLAY_NO_ATTRIBUTE, // sentinel
+            };
+            int32_t values[sizeof(attributes)/sizeof(attributes[0])];
 
-    int retireFenceFd = -1;
+            hwc_device->getDisplayAttributes(hwc_device, dpy, configs[i],
+                    attributes, values);
 
-    if (m_waitOnRetireFence) {
-        retireFenceFd = mlist[0]->retireFenceFd;
-        mlist[0]->retireFenceFd = -1;
+            fprintf(stderr, "Dpy %d Cfg %d (%d) VSYNC_PERIOD=%d, SIZE=(%d, %d), DPY=(%d, %d)\n",
+                    dpy, i, configs[i], values[0], values[1], values[2], values[3], values[4]);
+        }
     }
+}
 
-    if (m_syncBeforeSet) {
-        int acqFd = getFenceBufferFd(buffer);
-        if (acqFd >= 0) {
-            sync_wait(acqFd, -1);
-            close(acqFd);
+// contents for a single screen
+class HwComposerScreen_v11 {
+public:
+    enum Layer {
+        // Layers we use for composition
+        HWC_SCREEN_FRAMEBUFFER_LAYER = 0,
+        HWC_SCREEN_FRAMEBUFFER_TARGET_LAYER = 1,
+
+        // Number of layers we need to allocate space for
+        HWC_SCREEN_REQUIRED_LAYERS = 2,
+    };
+
+    HwComposerScreen_v11(hwc_composer_device_1_t *hwc_device, int id)
+        : hwc_device(hwc_device)
+        , id(id)
+        , hwc_list(nullptr)
+        , m_screen_width(0)
+        , m_screen_height(0)
+    {
+        size_t needed_size = sizeof(hwc_display_contents_1_t) +
+            HWC_SCREEN_REQUIRED_LAYERS * sizeof(hwc_layer_1_t);
+
+        hwc_list = (hwc_display_contents_1_t *) calloc(1, needed_size);
+
+        // Need to set this here, and not every time in relayout
+        hwc_list->numHwLayers = 2;
+        hwc_list->retireFenceFd = -1;
+#ifdef HWC_DEVICE_API_VERSION_1_3
+        hwc_list->outbuf = 0;
+        hwc_list->outbufAcquireFenceFd = -1;
+#endif
+    }
+
+    void update_size() {
+        get_screen_size(hwc_device, id, &m_screen_width, &m_screen_height);
+    }
+
+    bool relayout(int width, int height)
+    {
+        // Source rectangle of the desktop
+        const hwc_rect_t source_rect = {
+            0, 0, width, height
+        };
+
+        // Don't call getDisplayAttributes too often here since it can cause big slowdowns.
+        if (m_screen_width <= 0 || m_screen_height <= 0) {
```

Comment:
This will be still executed because not connected displays (external and virtual) will have m_screen_width/height = 0 until first connection. I wonder whether HwComposerScreen_v11 should be created on demand for external and virtual display. For now relayout is called for 3 displays even without external display connection.
Is it really needed to call relayout() on every frame? From my tests calling it after unblank display was enough but I never manged to get whole code working correctly on native portrait display device so take it with grain of salt.

Reply:
Does it have to be after unblank or is it enough if it's called on hotplug?

Reply:
I have added relayout(HwComposerBackend_v11->widht/height) in hotplug and in createWindow() after width/height are set and indeed both builtin and external display works on A31 based table I use for testing.